### PR TITLE
[TS] LPS-127109 Segments with both context and model criteria are not correctly evaluated with guest users

### PR DESCRIPTION
**Pull request by jorgediaz-lr**
https://issues.liferay.com/browse/LPS-127109

Originally sent in PR https://github.com/liferay-tango/liferay-portal/pull/837

Changes:

- I have merged all the changes in two commits.
- All the DefaultUser tests are splitin two: the normal one (with the SIGNED_IN parameter) and another xxxWithoutSignedInContext() without the SIGNED_IN parameter
- Renamed guestUser to defaultUser and sorted with the alphabetical name

### CODE REVIEW

**Comment by jorgediaz-lr** [comment]
ci:test:sf

**Comment by jorgediaz-lr** [comment]
ci:test:relevant

**Comment by jorgediaz-lr** [comment]
ci:test:tango

**Comment by darquesdev** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/provider/BaseSegmentsEntryProvider.java [hunk 1]
```diff
@@ -228,7 +228,24 @@ protected boolean isMember(
 		Criteria.Conjunction contextConjunction = getConjunction(
 			segmentsEntry, Criteria.Type.CONTEXT);
 
+		String modelFilterString = getFilterString(
+			segmentsEntry, Criteria.Type.MODEL);
+
 		if (context != null) {
+			boolean defaultUser = false;
+
+			if (context.containsKey(Context.SIGNED_IN) &&
```
Comment (darquesdev) [inline modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/provider/BaseSegmentsEntryProvider.java:237]: Is it better to directly calculate `defaultUser` value?

```java
boolean defaultUser = !GetterUtil.getBoolean(context.get(Context.SIGNED_IN), true);
```

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 2]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
 				segmentsEntryIds));
 	}
 
+	@Test
+	public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterionAndDefaultUser()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:335]: Please, define as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 3]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:338]: Again, define as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 4]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+			context);
+
+		Assert.assertArrayEquals(new long[0], segmentsEntryIds);
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterionAndDefaultUserWithoutSignedInContext()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:370]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 5]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:373]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 6]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+			context);
+
+		Assert.assertArrayEquals(new long[0], segmentsEntryIds);
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithContextCriterionOrModelCriterionAndDefaultUser()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:404]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 7]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:407]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 8]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+
+		Assert.assertArrayEquals(
+			new long[] {segmentsEntry.getSegmentsEntryId()}, segmentsEntryIds);
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithContextCriterionOrModelCriterionAndDefaultUserWithoutSignedInContext()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:440]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 9]
```diff
@@ -298,6 +328,146 @@ public void testGetSegmentsEntryIdsWithContextCriterionAndModelCriterion()
...
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:443]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 10]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
 	}
 
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndModelCriterionAndDefaultUser()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:657]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 11]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
 		Assert.assertArrayEquals(new long[0], segmentsEntryIds);
 	}
 
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndModelCriterionAndDefaultUser()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:654]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 12]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
+			context);
+
+		Assert.assertArrayEquals(new long[0], segmentsEntryIds);
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndModelCriterionAndDefaultUserWithoutSignedInContext()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:689]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 13]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:692]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 14]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionOrModelCriterionAndDefaultUser()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:726]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 15]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
+			context);
+
+		Assert.assertArrayEquals(new long[0], segmentsEntryIds);
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionOrModelCriterionAndDefaultUser()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:723]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 16]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionOrModelCriterionAndDefaultUserWithoutSignedInContext()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
+			TestPropsValues.getCompanyId());
+
+		User defaultUser = company.getDefaultUser();
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:761]: as used

File: modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java [hunk 17]
```diff
@@ -418,6 +647,144 @@ public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionAndMatchingMod
...
+			context);
+
+		Assert.assertArrayEquals(new long[0], segmentsEntryIds);
+	}
+
+	@Test
+	public void testGetSegmentsEntryIdsWithNonmatchingContextCriterionOrModelCriterionAndDefaultUserWithoutSignedInContext()
+		throws Exception {
+
+		Company company = _companyLocalService.getCompany(
```
Comment (darquesdev) [inline modules/apps/segments/segments-test/src/testIntegration/java/com/liferay/segments/provider/test/DefaultSegmentsEntryProviderTest.java:758]: as used

**Comment by darquesdev** [comment]
Hi @jorgediaz-lr, I've left some comments. Note: Try to use the same commit, forcing the push to the same branch. 
Apart from the comment, could you change the commits msg for something more meaningful? If not, just use "Add tests"

**Comment by jorgediaz-lr** [comment]
Hi @darquesdev

I have forced push the same branch with the requested changes.

Let me know in case you have any question.

Regards,
Jorge

**Comment by jorgediaz-lr** [comment]
ci:test:sf

**Comment by jorgediaz-lr** [comment]
ci:test:relevant

**Comment by jorgediaz-lr** [comment]
ci:test:tango

**Comment by darquesdev** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Review by darquesdev — APPROVED** [review]
Great PR @jorgediaz-lr, you can forward it's once It's green ;)

**Comment by jorgediaz-lr** [comment]
ci:test:relevant

**Comment by jorgediaz-lr** [comment]
ci:stop

**Comment by jorgediaz-lr** [comment]
ci:test:relevant

**Comment by jorgediaz-lr** [comment]
ci:test:sf

**Comment by jorgediaz-lr** [comment]
ci:test:relevant

**Comment by jorgediaz-lr** [comment]
ci:test:relevant

**Comment by jorgediaz-lr** [comment]
ci:test:tango

**Comment by jorgediaz-lr** [comment]
@darquesdev I have rebased the changes to latest master to force the executing of CI tests

**Comment by jorgediaz-lr** [comment]
ci:forward

**Comment by jorgediaz-lr** [comment]
ci:forward

**Comment by dgarciasarai** [comment]
ci:test:tango

**Pull request closed without merging**